In [3]:
import pickle
with open('models1.pkl', 'rb') as f:
  models = pickle.load(f)

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sklearn as sk
from sklearn.metrics import mean_absolute_percentage_error as mape
from sklearn.metrics import mean_absolute_error as mae
from sklearn.metrics import mean_squared_error as mse

In [5]:
with open('test.npy', 'rb') as f:
    X_test = np.load(f)
    y_test = np.load(f)

In [9]:
# linear regression
print('test R2_score = ', models[0].score(X_test, y_test))
print('test rmse_score = ', np.sqrt(mse(y_test, models[0].predict(X_test))))
print('test mape_score = ', mape(y_test, models[0].predict(X_test)))
print('test mae_score = ', mae(y_test, models[0].predict(X_test)))
print('(mape для y!=0):',mape(y_test[y_test!=0], models[0].predict(X_test)[y_test!=0]))

test R2_score =  0.7638522762409652
test rmse_score =  1.072284493316548
test mape_score =  99494980624846.48
test mae_score =  0.13278890084457187
(mape для y!=0): 22.86196157083573


In [12]:
# MLP regressor
print('test R2_score = ', models[1].score(X_test, y_test))
print('test rmse_score = ', np.sqrt(mse(y_test, models[1].predict(X_test))))
print('test mape_score = ', mape(y_test, models[1].predict(X_test)))
print('test mae_score = ', mae(y_test, models[1].predict(X_test)))
print('(mape для y!=0):',mape(y_test[y_test!=0], models[1].predict(X_test)[y_test!=0]))

test R2_score =  0.7844315226453266
test rmse_score =  1.0244971881004095
test mape_score =  28819444574364.25
test mae_score =  0.09425775305817256
(mape для y!=0): 7.521882858943411


In [13]:
# GradientBoostingRegressor
print('test R2_score = ', models[2].score(X_test, y_test))
print('test rmse_score = ', np.sqrt(mse(y_test, models[2].predict(X_test))))
print('test mape_score = ', mape(y_test, models[2].predict(X_test)))
print('test mae_score = ', mae(y_test, models[2].predict(X_test)))
print('(mape для y!=0):',mape(y_test[y_test!=0], models[2].predict(X_test)[y_test!=0]))

test R2_score =  0.7825550247756997
test rmse_score =  1.0289465890850418
test mape_score =  30619358834314.445
test mae_score =  0.0895138132817083
(mape для y!=0): 5.21730293656799


   За основну метрику виберемо MAE, так як вона найбільш чітко відображає реальне відхилення моделі, a MAPE не підходить для цих даних, тому що у вибірці занадто багато нульових значень і не можна виміряти якість прогнозів відносно абсолютних значень, замість неї можна використати метрику - коефіцієнт детермінації (якість прогнозів відносно дисперсії). Проведемо аналіз отриманих результатів. Досягнена точність 0.78 на тестовій вибірці. Це можна назвати гарним, за деякими канонами реалістичним результатом.
   
   Метрики:
        Корінь середньої квадратичної похибки трохи більше одиниці. Ця метрика трохи збільшує значимість малих похибок, тож ми не будемо її використовувати
        MAE середня сума абсолютної різниці між вірним і передбаченим показує хороші результати. Це говорить про те, що похибка у передбачених значеннях є невисокою. Рахуючи що ми працюємо з передбаченням реальних доходів, то ми вважаємо це хорошим результатом, тож обираємо саме цю метрику.
   
   Таким чином ми можемо спрогнозувати загальну картину наших потенційних доходів, що хоч і не буде 100% точною, але в загальному відображатиме, що нас очікує
   
   Серед трьох розглянутих моделей найбільш оптимальною, на думку команди є модель з використанням алгоритму GradientBoostingRegressor. Оскільки модель з лінійною регресією та MLP регресію мають менш задовільні показники MAE у порівнянні з обраною моделю.  